### PR TITLE
feat: implement tenants' mapping rules' migration

### LINKED ISSUE #25165: Camunda supports TenantMappingRule by implementing TenantMappingRuleMigrationHandler
### Description
Within scope of this issue, we assume that all available tenants are migrated and there is no need to create a tenant as part of migration. 
This issue needs [tenant migration task](https://github.com/camunda/camunda/issues/25130) be completed.

### PR DESCRIPTION
**Pull request by maryarm**
## Description

Migrate tenant mapping rules from Management identity to new identity.

## Related issues

closes #25165

### CODE REVIEW

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      final TenantMappingRule tenantMappingRule) {
+    try {
+      final var mappingKey =
+          mappingService.findOrCreateMapping(
+              tenantMappingRule.getName(),
+              tenantMappingRule.getClaimName(),
+              tenantMappingRule.getClaimValue());
+      for (final Tenant mappingTenant : tenantMappingRule.getAppliedTenants()) {
+        final var tenant =
+            tenantService.fetchOrCreateTenant(mappingTenant.tenantId(), mappingTenant.name());
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:66]: :question: Why is this needed? The tenant migration always runs first so all tenant should exist already. If it's just to fetch the tenant key, I'd try and make that explicit, i.e. split `fetchOrCreateTenant` in two, `fetch` and `create`.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              ? null
+              : tenantMappingRules.get(tenantMappingRules.size() - 1);
+    } while (lastTenantMappingRule != null);
+  }
+
+  private MigrationStatusUpdateRequest createTenantMappingRule(
+      final TenantMappingRule tenantMappingRule) {
+    try {
+      final var mappingKey =
+          mappingService.findOrCreateMapping(
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:60]: :question: What's the point in searching for the mapping first? As long as we handle conflicts (see my other comment) we should be able to always create it without searching first.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeConnector.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    final var properties = identityMigrationProperties.getZeebe();
+    return newZeebeClient(properties);
+  }
+
+  public ZeebeClient newZeebeClient(final ZeebeProperties zeebeProperties) {
+    final var gatewayAddress = getGatewayAddress(zeebeProperties);
+    final ZeebeClientBuilder builder =
+        ZeebeClient.newClientBuilder()
+            .gatewayAddress(gatewayAddress)
+            .restAddress(URI.create("http://0.0.0.0:8081"))
```
[the remark this answers is not in the document]
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeConnector.java:39]: Ah, yes, it leaked from my local changes.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 1 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:66]: yes, but I wanted to reuse it for TenantMigrationHandler when we create the tenant and also with this we are more flexible in ordering the handlers.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+              tenantMappingRule.getName(),
+              tenantMappingRule.getClaimName(),
+              tenantMappingRule.getClaimValue());
```
[the remark this answers is not in the document]
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:63]: because here we need the mapping key for next step (assign mapping to tenant) to be available, so here conflict is a error and we need to retry. ( I have to add that in this case it's very unlikely to have a conflict because a search query is executed before and it's just for very rare case that the mapping is created but is not  published to elastic yet.)

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/MigrationHandler.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.migration.identity;
+
+public interface MigrationHandler {
+
+  void migrate();
+
+  default boolean isConflictError(final Exception e) {
+    return e.getMessage().contains("Failed with code 409: 'Conflict'");
```
[the remark this answers is not in the document]
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/MigrationHandler.java:15]: I'll check again but as I remember the exception is wrapped in another exception class that don't expose the status and checking the message is safer.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 2 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:60]: I need the mappingKey for next step but in case of conflict I can get the mappingKey.

File: dist/src/main/resources/application.properties [hunk 6]
```diff
@@ -6,7 +6,7 @@ logging.register-shutdown-hook=true
 # This sever is enabled even for a standalone broker without any embedded gateway, in which case
 # there will simply be no endpoints available.
 server.address=0.0.0.0
-server.port=8080
+server.port=8081
```
Comment (lenaschoenburg) [inline dist/src/main/resources/application.properties:9]: :x: this looks a change from local testing

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 1 above)
Comment (lenaschoenburg) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:66]: Since the refresh interval of ES indices is 1 second, I think there's a good chance that we will first create all tenants and then try to create them again here because we can't find them in ES yet. That would lead to CONFLICT errors which we don't handle here so the migration app would just fail.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lenaschoenburg) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:63]: But the retry would only occur because the migration app crashes, right? I think we should try to handle this explicitly and not rely on the crash.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeProperties.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.migration.identity.config;
+
+public class ZeebeProperties {
+
+  private String brokerContactPoint;
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeProperties.java:12]: :wrench: Since this is an entirely new configuration (under `camunda.migration.identity.zeebe`) I think we can ignore the backwards compatibility stuff around `brokerContactPoint` and only use `gatewayAddress` directly. So `ZeebeConnector#getGatewayAddress` can be removed too.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:63]: Actually not, retry will happens automatically on next loop during fetch mapping rules. The endpoints in identity designed to return failed or not already migrated errors. To be precise it was designed like this and I applied a fix in next PR to actually achieve it.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 1 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:66]: Conflict here is managed as an error status and we don't throw it to the caller, so in next loop for fetching the tenants we have them and can try. By the way do you think it's a good idea to have a pause (sleep) for a few seconds when we first experienced this error? so we have less conflict error.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeProperties.java (hunk 7 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/config/ZeebeProperties.java:12]: I'll fix it in the next PR that will be ready by today.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java (hunk 1 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java:66]: In my new PR, I have separated the `create` and `fetch` methods and I hope it prevent congesting zeebe with a lot of conflict requests.

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/service/MappingService.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      final ZeebeClient zeebeClient,
+      final Authentication.Builder servicesAuthenticationBuilder) {
+    this.mappingServices =
+        mappingServices.withAuthentication(servicesAuthenticationBuilder.build());
+    this.zeebeClient = zeebeClient;
+  }
+
+  public Long findOrCreateUserWithUsername(final String username) {
+    return findOrCreateMapping(username + "_mapping", USERNAME_CLAIM, username);
+  }
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/main/java/io/camunda/migration/identity/service/MappingService.java:35]: :question: unused?

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/dto/MappingRule.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  public enum MappingRuleType {
+    ROLE,
+    TENANT,
+    GROUP;
+  }
+
+  public enum Operator {
+    CONTAINS,
+    EQUALS;
+  }
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/main/java/io/camunda/migration/identity/dto/MappingRule.java:82]: :question: I guess this is also not yet supported, right? Perhaps we should throw an exception when encountering `CONTAINS` until it is supported?

File: migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  @Mock private TenantService tenantService;
+  @Mock private MappingService mappingService;
+
+  private final ManagementIdentityTransformer managementIdentityTransformer =
+      new ManagementIdentityTransformer();
+  private TenantMappingRuleMigrationHandler migrationHandler;
+
+  @BeforeEach
+  void setUp() throws Exception {
+    MockitoAnnotations.openMocks(this).close();
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java:53]: :question: What is this doing?

File: migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    // when
+    migrationHandler.migrate();
+
+    // then
+    verify(managementIdentityClient, times(2)).fetchTenantMappingRules(anyInt());
+    verify(tenantService, never()).fetch(any(), any());
+    verify(mappingService, times(2)).findOrCreateMapping(any(), any(), any());
+    verify(managementIdentityClient, times(2))
+        .updateMigrationStatus(migrationStatusCaptor.capture());
```
Comment (lenaschoenburg) [inline migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java:94]: :wrench: No argument captor needed, you can just verify the arguments directly here

**Review by lenaschoenburg — APPROVED** [review]

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/dto/MappingRule.java (hunk 9 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/dto/MappingRule.java:82]: In new identity both EQUALS and CONTAINS are same

File: migration/identity-migration/src/main/java/io/camunda/migration/identity/service/MappingService.java (hunk 8 above)
Comment (maryarm) [reply migration/identity-migration/src/main/java/io/camunda/migration/identity/service/MappingService.java:35]: It leaked from next PR. ;)

File: migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java (hunk 10 above)
Comment (maryarm) [reply migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java:53]: It's for creating @Mock instances. If you know a better approach, let me know and I apply in next tests.

Comment (lenaschoenburg) [reply migration/identity-migration/src/test/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandlerTest.java:53]: Ah okay, I think you could also just annotate the test class with ` @ExtendWith(MockitoExtension. class)`
